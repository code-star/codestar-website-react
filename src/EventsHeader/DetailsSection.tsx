import React, { SFC } from 'react';
import { Typography, Button } from '@material-ui/core';
import { translate, TranslationFunction } from 'react-i18next';
import Container from '../Container/Container';
import Section from '../Section/Section';
import compose from 'recompose/compose';

// TODO improve types by replacing "any"
interface IDetailsSectionPropsInner {
	t: TranslationFunction;
	mEvent: any;
	formattedDate: string;
	descriptionElem: any;
}

interface IDetailsSectionPropsOuter {
	mEvent: any;
	formattedDate: string;
	descriptionElem: any;
}

export const DetailsSection: SFC<IDetailsSectionPropsInner> = ({
	t,
	mEvent,
	formattedDate,
	descriptionElem,
}) => {
	return (
		<Section scrollname="event-details" className="bg-white">
			<Container center>
				<div className="row">
					<div className="col-12">
						<Typography align="center" variant="title">
							{mEvent.name}
						</Typography>
						<Typography gutterBottom align="center" variant="subheading">
							{formattedDate}
						</Typography>
					</div>
				</div>
				<div className="row">
					<div className="col-12 col-md-8">{descriptionElem}</div>
					<div className="col-12 col-md-4">
						<img
							src={mEvent.coverUrl}
							alt={`Artistic background with text "${mEvent.name}"`}
							style={{ width: '100%' }}
						/>
						<Button
							color="primary"
							variant="raised"
							href={mEvent.link}
							className="mt-1"
						>
							{t('SIGN_UP')}
						</Button>
					</div>
				</div>
			</Container>
		</Section>
	);
};

export default compose<IDetailsSectionPropsInner, IDetailsSectionPropsOuter>(
	translate(['events'], { wait: true })
)(DetailsSection);
