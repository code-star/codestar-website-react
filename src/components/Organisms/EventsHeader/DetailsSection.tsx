import React, { SFC, ReactNode } from 'react';
import { Typography, Button } from '@material-ui/core';
import { translate, TranslationFunction } from 'react-i18next';
import Container from '../../../Container/Container';
import Section from '../../Molecules/Section/Section';
import compose from 'recompose/compose';
import TweetList from '../../Molecules/List/TweetList';
import {
  IMeetupEvent,
  ITweet,
} from '../../../containers/EventsContainer/EventsContainer.interfaces';

interface IDetailsSectionPropsInner {
  t: TranslationFunction;
}

interface IDetailsSectionPropsOuter {
  mEvent: IMeetupEvent;
  formattedDate: string;
  descriptionElem: ReactNode;
  tweets: ITweet[];
}

export const DetailsSection: SFC<
  IDetailsSectionPropsInner & IDetailsSectionPropsOuter
> = ({ t, mEvent, formattedDate, descriptionElem, tweets }) => {
  return (
    <Section scrollname="event-details" className="bg-white">
      <Container center>
        <div className="row">
          <div className="col-12 col-md-8">
            <Typography align="center" variant="h6">
              {mEvent.name}
            </Typography>
            <Typography gutterBottom align="center" variant="subtitle1">
              {formattedDate}
            </Typography>
            {descriptionElem}
          </div>
          <div className="col-12 col-md-4">
            <TweetList
              tweets={tweets}
              eventDate={formattedDate}
              eventImage={mEvent.coverUrl}
              eventName={mEvent.name}
            >
              <Button color="primary" href={mEvent.link}>
                {t('SIGN_UP')}
              </Button>
            </TweetList>
          </div>
        </div>
      </Container>
    </Section>
  );
};

export default compose<
  IDetailsSectionPropsInner & IDetailsSectionPropsOuter,
  IDetailsSectionPropsOuter
>(translate(['events'], { wait: true }))(DetailsSection);
