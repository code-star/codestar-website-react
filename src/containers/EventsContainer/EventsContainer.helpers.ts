import { get } from 'lodash'

export function convertEventResponseToModel(withDescription = false) {
  return (mEvent: any) => {
    const fallbackImage = 'https://res.cloudinary.com/codestar/image/upload/v1532409289/codestar.nl/meetup/codestar-night-logo.jpg'
    const result = {
      name: mEvent.name,
      time: mEvent.time,
      link: mEvent.link,
      coverUrl: get(mEvent, 'featured_photo.photo_link', fallbackImage),
      withDescription,
    }
    if (withDescription) {
      Object.assign(result, {
        description: mEvent.description,
      })
    }
    return result
  }
}
