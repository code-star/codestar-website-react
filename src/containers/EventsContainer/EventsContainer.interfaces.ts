import { VideoItem } from './fetchYouTubePlaylist';

export interface IMeetupEvent {
  coverUrl: string;
  link: string;
  name: string;
  time: number;
  withDescription: boolean;
  description?: string;
}

export interface IEventsContainerState {
  nextMeetupEvents: IMeetupEvent[];
  loadingNextMeetupEvent: boolean;
  noNextMeetupEvent: boolean;
  pastMeetupEvents: IMeetupEvent[];
  recentTweets: any[];
  videos: VideoItem[];
}

export interface ITwitterUser {
  name: string;
  profile_image_url_https: string;
  screen_name: string;
}

export interface ITweet {
  id: number;
  id_str: string;
  user: ITwitterUser;
  created_at: string;
  text: string;
}
