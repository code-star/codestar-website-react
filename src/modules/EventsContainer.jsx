import React, { Component } from 'react';
import Events from '../components/Events';
import { jsonp } from '../util';
import _ from 'lodash';

// Meetup API test console: https://secure.meetup.com/meetup_api/console/?path=/:urlname/events
// page=3 = number of results to return in a page, only need the first 3 results
const GET_UPCOMING_EVENTS_URL =
	'https://api.meetup.com/Code-Star-Night/events?photo-host=public&page=3&sig_id=226887185&status=upcoming&fields=featured_photo&sig=d035ab8a8f521cbb4ef14eaff79a55f23c3d25eb';

const GET_PAST_EVENTS_URL =
	'https://api.meetup.com/Code-Star-Night/events?desc=true&photo-host=public&sig_id=226887185&status=past&fields=featured_photo&sig=a60e663f0904424f80fda3b00bf31f315889231c';

/* TODO Convert JSONP to Fetch+Serverless
This url gets the event details, but is signed for one specific instance: https://api.meetup.com/Code-Star-Night/events/248958146?photo-host=public&sig_id=226887185&fields=featured_photo&sig=c634269c86bda35c0762874a490d219faba6365e
Can use RxJS with JSONP? To combine streams?
 */

function convertEventResponseToModel(withDescription = false) {
	return function(mEvent) {
		const fallbackImage =
			'https://res.cloudinary.com/codestar/image/upload/v1532409289/codestar.nl/meetup/codestar-night-logo.jpg';
		const result = {
			name: mEvent.name,
			time: mEvent.time,
			link: mEvent.link,
			coverUrl: _.get(mEvent, 'featured_photo.photo_link', fallbackImage),
			withDescription,
		};
		if (withDescription) {
			Object.assign(result, {
				description: mEvent.description,
			});
		}
		return result;
	};
}

export default class EventsContainer extends Component {
	state = {
		nextEvent: {
			loading: true,
			event: null,
			noEvent: false,
		},
		pastEvents: [],
	};

	componentDidMount() {
		this.fetchEvents();
	}

	async fetchEvents() {
		/* Meetup API only allows JSONP for client-side, non authenticated, api key signed GET requests.
		   must use JSONP conform https://github.com/meetup/api/issues/211
		   Fetch API does not support JSONP. no-cors mode creates an opaque response without data.
		*/
		try {
			const response = await jsonp(GET_UPCOMING_EVENTS_URL);
			const result = _.head(
				response.data.map(convertEventResponseToModel(true))
			);
			const nextEvent = {
				loading: false,
				mEvent: result ? result : null,
				noEvent: !result,
			};
			this.setState({ nextEvent });
		} catch (err) {
			this.setState({
				loading: false,
				event: null,
				noEvent: true,
			});
		}

		try {
			const response = await jsonp(GET_PAST_EVENTS_URL);
			const result = response.data.map(convertEventResponseToModel());
			this.setState({ pastEvents: result });
		} catch (err) {
			this.setState({ pastEvents: [] });
		}
	}

	render() {
		return (
			<Events
				nextEvent={this.state.nextEvent}
				pastEvents={this.state.pastEvents}
			/>
		);
	}
}
