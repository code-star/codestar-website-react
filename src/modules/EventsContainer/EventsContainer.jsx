import React, { Component } from 'react';
import Events from '../../components/Molecules/List/Events';
import _ from 'lodash';
import {
  getCachedUpcomingEvents,
  getCachedPastEvents,
  getCachedRecentTweets,
} from '../../eventsService';

function convertEventResponseToModel(withDescription = false) {
  return function(mEvent) {
    const fallbackImage =
      'https://res.cloudinary.com/codestar/image/upload/v1532409289/codestar.nl/meetup/codestar-night-logo.jpg';
    const result = {
      name: mEvent.name,
      time: mEvent.time,
      link: mEvent.link,
      coverUrl: _.get(mEvent, 'featured_photo.photo_link', fallbackImage),
      withDescription,
    };
    if (withDescription) {
      Object.assign(result, {
        description: mEvent.description,
      });
    }
    return result;
  };
}

export default class EventsContainer extends Component {
  state = {
    nextMeetupEvents: [],
    loadingNextMeetupEvent: true,
    noNextMeetupEvent: false,
    pastMeetupEvents: [],
    recentTweets: [],
  };

  componentDidMount() {
    this.fetchEvents();
    this.fetchRecentTweets();
  }

  async fetchEvents() {
    try {
      const response = await getCachedUpcomingEvents();
      const nextMeetupEvents = response.map(convertEventResponseToModel(true));
      this.setState({
        nextMeetupEvents,
        loadingNextMeetupEvent: false,
        noNextMeetupEvent: !(nextMeetupEvents && nextMeetupEvents.length > 0),
      });
    } catch (err) {
      this.setState({
        nextMeetupEvents: null,
        loadingNextMeetupEvent: false,
        noNextMeetupEvent: true,
      });
    }

    try {
      const response = await getCachedPastEvents();
      const pastMeetupEvents = response.map(convertEventResponseToModel());
      this.setState({ pastMeetupEvents });
    } catch (err) {
      this.setState({ pastMeetupEvents: [] });
    }
  }

  async fetchRecentTweets() {
    try {
      const response = await getCachedRecentTweets();
      this.setState({ recentTweets: response });
    } catch (err) {
      this.setState({ recentTweets: [] });
    }
  }

  render() {
    const {
      nextMeetupEvents,
      noNextMeetupEvent,
      pastMeetupEvents,
      recentTweets,
    } = this.state;

    return (
      <Events
        nextMeetupEvents={nextMeetupEvents}
        noNextMeetupEvent={noNextMeetupEvent}
        pastMeetupEvents={pastMeetupEvents}
        recentTweets={recentTweets}
      />
    );
  }
}
