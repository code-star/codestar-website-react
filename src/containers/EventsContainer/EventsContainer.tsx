import React, { Component } from 'react'
import Events from '../../components/Molecules/List/Events'
import { getCachedUpcomingEvents, getCachedPastEvents, getCachedRecentTweets } from '../../eventsService'
import { IEventsContainerState } from './EventsContainer.interfaces'
import { convertEventResponseToModel } from './EventsContainer.helpers'
import { fetchYouTubePlaylist } from './fetchYouTubePlaylist'
import { YOUTUBE_API_KEY, YOUTUBE_CODESTAR_PLAYLIST_ID } from '../../constants'

export default class EventsContainer extends Component<{}, IEventsContainerState> {
  public state: IEventsContainerState = {
    nextMeetupEvents: [],
    loadingNextMeetupEvent: true,
    noNextMeetupEvent: false,
    pastMeetupEvents: [],
    recentTweets: [],
    videos: [],
  }

  public componentDidMount() {
    this.fetchEvents()
    this.fetchRecentTweets()
    this.fetchVideos()
  }

  public render() {
    const { nextMeetupEvents, noNextMeetupEvent, pastMeetupEvents, recentTweets, videos }: IEventsContainerState = this.state

    return (
      <Events
        nextMeetupEvents={nextMeetupEvents}
        noNextMeetupEvent={noNextMeetupEvent}
        pastMeetupEvents={pastMeetupEvents}
        recentTweets={recentTweets}
        videos={videos}
      />
    )
  }

  private filterRecentTweets(tweets: any) {
    return tweets.filter((tweet: any) => !tweet.text.includes('RT ') && !tweet.text.startsWith('@'))
  }

  private async fetchEvents() {
    try {
      const response: any = await getCachedUpcomingEvents()
      const nextMeetupEvents = response.map(convertEventResponseToModel(true))
      this.setState({
        nextMeetupEvents,
        loadingNextMeetupEvent: false,
        noNextMeetupEvent: !(nextMeetupEvents && nextMeetupEvents.length > 0),
      })
    } catch (err) {
      this.setState({
        nextMeetupEvents: [],
        loadingNextMeetupEvent: false,
        noNextMeetupEvent: true,
      })
    }

    try {
      const response: any = await getCachedPastEvents()
      const pastMeetupEvents = response.map(convertEventResponseToModel())
      this.setState({ pastMeetupEvents })
    } catch (err) {
      this.setState({ pastMeetupEvents: [] })
    }
  }

  private async fetchRecentTweets() {
    try {
      const response: any = await getCachedRecentTweets()

      const recentTweets = this.filterRecentTweets(response)

      this.setState({ recentTweets })
    } catch (err) {
      this.setState({ recentTweets: [] })
    }
  }

  private async fetchVideos() {
    const videos = await fetchYouTubePlaylist(YOUTUBE_API_KEY, YOUTUBE_CODESTAR_PLAYLIST_ID)
    this.setState({
      videos,
    })
  }
}
