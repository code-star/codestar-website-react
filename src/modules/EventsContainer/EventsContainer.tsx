import React, { Component } from 'react';
import Events from '../../components/Molecules/List/Events';
import {
  getCachedUpcomingEvents,
  getCachedPastEvents,
  getCachedRecentTweets,
} from '../../eventsService';
import { IEventsContainerState } from './EventsContainer.interfaces';
import { convertEventResponseToModel } from './EventsContainer.helpers';

export default class EventsContainer extends Component<
  {},
  IEventsContainerState
> {
  public state: IEventsContainerState = {
    nextMeetupEvents: [],
    loadingNextMeetupEvent: true,
    noNextMeetupEvent: false,
    pastMeetupEvents: [],
    recentTweets: [],
  };

  public componentDidMount() {
    this.fetchEvents();
    this.fetchRecentTweets();
  }

  public render() {
    return (
      <Events
        nextMeetupEvents={this.state.nextMeetupEvents}
        noNextMeetupEvent={this.state.noNextMeetupEvent}
        pastMeetupEvents={this.state.pastMeetupEvents}
        recentTweets={this.state.recentTweets}
      />
    );
  }

  private async fetchEvents() {
    try {
      const response: any = await getCachedUpcomingEvents();
      const nextMeetupEvents = response.map(convertEventResponseToModel(true));
      this.setState({
        nextMeetupEvents,
        loadingNextMeetupEvent: false,
        noNextMeetupEvent: !(nextMeetupEvents && nextMeetupEvents.length > 0),
      });
    } catch (err) {
      this.setState({
        nextMeetupEvents: [],
        loadingNextMeetupEvent: false,
        noNextMeetupEvent: true,
      });
    }

    try {
      const response: any = await getCachedPastEvents();
      const pastMeetupEvents = response.map(convertEventResponseToModel());
      this.setState({ pastMeetupEvents });
    } catch (err) {
      this.setState({ pastMeetupEvents: [] });
    }
  }

  private async fetchRecentTweets() {
    try {
      const response: any = await getCachedRecentTweets();
      this.setState({ recentTweets: response });
    } catch (err) {
      this.setState({ recentTweets: [] });
    }
  }
}
