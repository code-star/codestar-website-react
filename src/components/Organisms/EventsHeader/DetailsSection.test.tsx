import React from 'react';
import renderer from 'react-test-renderer';
import { DetailsSection } from './DetailsSection';
import { IMeetupEvent } from '../../../containers/EventsContainer/EventsContainer.interfaces';

// Mock for "t", the translate function from react-i18next
const t = (text: string | string[]) => text;

it('renders the details for the Events header', () => {
  const meetupEvent: IMeetupEvent = {
    coverUrl: 'foo.png',
    link: 'https://www.codestar.nl',
    name: 'Data Oriented Design with Maxim Zaks',
    time: 123,
    withDescription: false,
  };
  const formattedDate = '26 september 2018';
  const comp = renderer
    .create(
      <DetailsSection
        t={t}
        mEvent={meetupEvent}
        formattedDate={formattedDate}
        descriptionElem={'descriptionElem'}
        tweets={[]}
      >
        <button />
      </DetailsSection>
    )
    .toJSON();
  expect(comp).toMatchSnapshot();
});
