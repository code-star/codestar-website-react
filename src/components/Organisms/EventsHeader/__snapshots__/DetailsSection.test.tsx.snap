// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the details for the Events header 1`] = `
<section
  className="
			bg-white
			WithTheme(WithWidth(Section))-section"
  name="event-details"
>
  <div
    className="WithTheme(WithWidth(Section))-element"
    name="event-details"
  />
  <div
    className="
			container
			
			
			
			
			
			WithTheme(WithWidth(Container))-containerCenter
		"
  >
    <div
      className="row"
    >
      <div
        className="col-12 col-md-8"
      >
        <h2
          className="MuiTypography-root MuiTypography-title MuiTypography-alignCenter"
        >
          Data Oriented Design with Maxim Zaks
        </h2>
        <h3
          className="MuiTypography-root MuiTypography-subheading MuiTypography-gutterBottom MuiTypography-alignCenter"
        >
          26 september 2018
        </h3>
        descriptionElem
      </div>
      <div
        className="col-12 col-md-4"
      >
        <div
          className="MuiPaper-root MuiPaper-elevation8 MuiPaper-rounded MuiCard-root card"
        >
          <div
            className="MuiCardMedia-root media"
            style={
              Object {
                "backgroundImage": "url(\\"foo.png\\")",
              }
            }
            title="Data Oriented Design with Maxim Zaks"
          />
          <div
            className="MuiCardContent-root"
          />
          <div
            className="MuiCardActions-root"
          >
            <a
              className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-flat MuiButton-flatPrimary MuiCardActions-action"
              href="https://www.codestar.nl"
              onBlur={[Function]}
              onContextMenu={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              role="button"
              tabIndex="0"
            >
              <span
                className="MuiButton-label"
              >
                SIGN_UP
              </span>
              <span
                className="MuiTouchRipple-root"
              />
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;
